### PR TITLE
Documentation for ECS internals

### LINKED ISSUE #3372: Reflection allows users to clone arbitrary reflected components, but Reflect does not have a Clone bound
## Problem

1. [ReflectComponent::copy_component](https://github.com/bevyengine/bevy/blob/11f52b8c72fc3a568e8bb4a4cd1f3eb025ac2e13/crates/bevy_ecs/src/reflect.rs#L65) allows users to copy any arbitrary component data to different entities, and across worlds, as long as it is reflect-able.
2. Neither `Reflect` nor `Component` has a `Clone` bound, and so we can sneakily clone components that are not expecting to be cloned. I'm pretty sure this is ultimately done by just copying their raw memory.
3. Components are effectively cloned in a way that bypasses the `Clone` method, which can cause issues with things like ref-counting (used for `Handle`).

## Possible Solutions

1. Give `Reflect` and/or `Component` a `Clone` bound. This fixes problem 2, but not problem 3, and may be painful in niche use cases (although I have not seen any concrete use cases for components that cannot be cloned raised, see #1515).

Furthermore, not all fields of a component should be reflected. Cloning components with some cloneable field and defaulting the others is a reasonable strategy here, but does not permit a `Clone` bound.

2. Use the appropriate `Clone` methods when they exist.

3. Leave the existing behavior in place, and very clearly document what's going on and how to use it safely.

### PR DESCRIPTION
**Pull request by alice-i-cecile**
# Objective

- The internals of Bevy ECS are very intimidating.
- They are also poorly documented.
- Nevertheless, they are both important to not mess up and very intriguing for a certain class of new contributor / user.
- Closes #3372.

## Solution

- Document *all* the things.
- Also improve existing documentation within bevy_ecs where I see the opportunity.

## Status

This is a draft PR, so y'all can complain about my explanations as I work.

Unless there are strong objections, I'd like to tackle most of the crate in one pass, so then I can ensure that things aren't missed and the concepts are explained coherently.

### CODE REVIEW

File: crates/bevy_ecs/src/archetype.rs [hunk 1]
```diff
@@ -14,6 +14,7 @@ use std::{
     ops::{Index, IndexMut},
 };
 
+/// The unique identifier of an [`Archetype`]
```
Comment (DJMcNab) [inline crates/bevy_ecs/src/archetype.rs:17]: I think these are only unique, per `World`, as they are used for indexing.

File: crates/bevy_ecs/src/archetype.rs [hunk 2]
```diff
@@ -33,16 +34,22 @@ impl ArchetypeId {
     }
 }
 
+/// Records whether a component was modified by addition or mutation
```
Comment (DJMcNab) [inline crates/bevy_ecs/src/archetype.rs:37]: (Specifically in the case of adding a bundle to an archetype, where adding means that it was not in the archetype, and mutating means it was)

File: crates/bevy_ecs/src/archetype.rs [hunk 3]
```diff
@@ -116,6 +123,9 @@ pub(crate) struct ArchetypeComponentInfo {
     pub(crate) archetype_component_id: ArchetypeComponentId,
 }
 
+/// A unique set of components stored within the [`World`](crate::world::World)
```
Comment (DJMcNab) [inline crates/bevy_ecs/src/archetype.rs:126]: With sparse set components, these can contain many sets of component, although each set is uniquely assigned to a single archetype.

**Comment by Nilirad** [comment]
> Document *all* the things.

~~Start by putting `#![warn(missing_docs)]`~~

**Comment by alice-i-cecile** [comment]
> Start by putting #![warn(missing_docs)]

~~I think you mean #![forbid(missing_docs)]~~. One day!

**Comment by cart** [comment]
> Document all the things

I propose documenting sections of bevy_ecs in multiple prs instead of all at once to make reviewing + merging easier. I won't block on that, but it has my preference.

**Comment by alice-i-cecile** [comment]
> I propose documenting sections of bevy_ecs in multiple prs instead of all at once to make reviewing + merging easier. I won't block on that, but it has my preference.

Sounds good @cart. I'm probably going to work in this PR, then close it out and cherrypick the commits into a dozen mini-PRs.

**Comment by alice-i-cecile** [comment]
Alright, I think that's about all of it for a first pass. I'm going to set this down for a couple of days, and then come back to split the PRs and close this out when it's done.

Most of these docs could do with being expanded with context and examples, but that will be better handled in small PRs that experts can quickly review.

**Comment by alice-i-cecile** [comment]
Split into #3384, #3385, #3386, #3387, #3388, #3389, #3390 and #3391 😂  Closing this out now.

**Pull request closed without merging**